llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,455 @@
+
+static SmallVector<MCPhysReg, 16>
+getBBForbiddenReplacements(const MachineBasicBlock *MBB, const VirtRegMap &VRM,
+                           const MachineRegisterInfo &MRI,
+                           const AIEBaseRegisterInfo *TRI);
+
+static bool isUndefReserved(
+    const MachineOperand &Def, const MachineInstr &MI,
+    const MachineBasicBlock &MBB,
+    std::map<unsigned, SmallVector<const MachineInstr *, 4>> &UndefPartReuse,
+    const VirtRegMap &VRM);
+
+static bool isNotLastDef(const MachineOperand &Def,
+                         const MachineInstr &DefInstr,
+                         const MachineBasicBlock &MBB);
+
+static bool reuseSubRegs(SmallVector<const MachineInstr *, 4> &OrigInstr,
+                         SmallVector<const MachineInstr *, 4> &OtherInstr);
+
+static bool containEqualSubRegDefs(const MachineInstr *InstrA,
+                                   const MachineInstr *InstrB);
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  auto &TRI =
+      *static_cast<const AIEBaseRegisterInfo *>(MRI.getTargetRegisterInfo());
+  VirtRegMap &VRM = getAnalysis<VirtRegMap>();
+  LiveRegMatrix &LRM = getAnalysis<LiveRegMatrix>();
+  LiveIntervals &LIS = getAnalysis<LiveIntervals>();
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<const MachineBasicBlock *, 4> LoopBBs =
+      AIELoopUtils::getLoopBBs(MLI);
+
+  if (LoopBBs.empty())
+    return false;
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  bool Modified = renameMBBPhysRegs(LoopBBs, MF, VRM, LRM, MRI, LIS, TRI);
+  return Modified;
+}
+
+static bool
+renameMBBPhysRegs(const SmallVector<const MachineBasicBlock *, 4> &LoopBBs,
+                  const MachineFunction &MF, VirtRegMap &VRM,
+                  LiveRegMatrix &LRM, const MachineRegisterInfo &MRI,
+                  const LiveIntervals &LIS, const AIEBaseRegisterInfo &TRI) {
+  bool Modified = false;
+  for (const MachineBasicBlock *MBB : LoopBBs) {
+    LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+               dbgs() << "\n");
+    SmallVector<MCPhysReg, 16> UsedPhyRegs;
+    SmallVector<MCPhysReg, 16> UsedVirtRegs;
+    SmallVector<MCPhysReg, 16> AlreadyReplaced;
+    SmallVector<std::pair<MCRegister, MCRegister>, 8> ReplacementRegs;
+
+    SmallVector<MCPhysReg, 16> ForbiddenPhysRegs =
+        getBBForbiddenReplacements(MBB, VRM, MRI, &TRI);
+
+    std::map<unsigned, SmallVector<const MachineInstr *, 4>> UndefPartReuse;
+
+    for (const MachineInstr &MI : *MBB) {
+
+      if (isCopyElimination(MI, VRM))
+        continue;
+
+      for (const MachineOperand &Op : MI.defs()) {
+        Register Reg = Op.getReg();
+        if (!Op.getReg().isVirtual())
+          continue;
+        if (VRM.hasRequiredPhys(Op.getReg()))
+          continue;
+        if (Op.isTied())
```

Comment:
I think we need to check `isRenamable()` as well. I would then assume we do not need to check `isTied()`

Reply:
`isRenamable()` is only possible on physical registers. Since the pass only looks at virtual registers, this check is not possible and relevant

Reply:
Can you check how `isRenamable` is inferred then? I'd like to make sure we are not missing a special case here.

Reply:
ARM has some register allocation constraints, that are not contained in operand register classes.
This function is a helper function for these cases.